### DISCUSSION TITLE
Welcome to LVSolutionBuilder Discussions!

### DISCUSSION THREAD

**Discussion by jovianarts** [jovianarts/LVSolutionBuilder#1] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we build together 💪.

**Comment by joviancw** [comment]
@jovianarts I am getting requests about this repository. Why is https://www.vipm.io/package/lv_solution_builder/ 
Why is this project under my VIPM account. I have nothing to do with this repo.

**Reply by jovianarts** [reply]
@joviancw I don't actually know. I'm assuming a clerical error, that someone started typing "jovian", picked the name that was recommended, and went with that. I had nothing to do with getting this project added to VIPM and knew nothing of that action until your message. You're welcome to contact whoever to get that addressed.

**Reply by jovianarts** [reply]
@joviancw I'm also going to try to see what I can do on my end. Maybe one of us will get lucky.

**Reply by jovianarts** [reply]
@joviancw , I contacted JKI and their response was excellent and immediate. They reassured me that they take inquiries like mine very seriously and backed up their words with action. The page you posted above is now taken down.

**Comment by labview-mike** [comment]
@jovianarts 

Hi Phil,

First, thank you for creating such a useful tool. It's been essential for our company's extensive use of PPLs. We have over 100 PPLs in our common reuse libraries. Before switching to Solution Builder, we were using Solution Explorer where the source of each "PPL layer" was linked to all the PPLs below it. That made editing much harder.

That being said, I'm running into a couple problems - probably because our use case is so complicated. I think we're at the bleeding edge of what the IDE is happy with. If you have any troubleshooting ideas or other suggestions, would really appreciate your time.

Issue 1: **IDE crashes when attempting to build the entire source repo**

When I run Solution Builder on our common project (with all ~120 libraries in it), the IDE will crash before it reaches the final library. It always happens toward the end, when we've gotten more than 90% of the way through.

If I add a "Save All" node after the PPL replacements for each library, then everything builds all the way to the end. This works, but then I have to revert to the previous commit after every build. Feels a little hacky.

I'm not sure what limitation we're running into here, maybe a cache is growing too large/unwieldy for the IDE to handle?

Issue 2: **Solution Builder sometimes claims there's a circular dependency where none exists**

A couple times, Solution Builder has claimed there's a circular dependency and refuses to build. I know for sure the dependency is absent, for the following reasons:

1) The project for each library doesn't reference the other in Dependencies view.
2) They successfully build into PPLs using either a manual build process or MGI Solution Explorer.

I've gotten around this by building in two steps, where the first step is everything up until the first "claimed" member of the circular pair and the second step is everything after. I don't know if it's relevant, but the last two libraries can be built at the very end anyway, as they're in the last layer of dependencies.

I tried digging into the code to see how I can replace the generated dependencies list with my own (using code I stole from Composed Systems / Jon McBee's dependency viewer). But I haven't been able to identify a good place to do that.

**Reply by jovianarts** [reply]
@labview-mike , thanks for the feedback and the questions!

Let me start by asking a clarifying question. When you say *"I've gotten around this by building in two steps"*, how are you accomplishing this? Are you using the `AddPackedLib` feature when building for step 2 or some other way to accomplish this?

The reason that I'm asking is because I was going to suggest building in 2 passes to bypass Issue 1. You likely noticed in the code that we keep references to updated VIs in open memory while the build occurs so that all callers also get updated. In your case, this could mean keeping a lot of VIs in memory, which could be difficult for LV to handle in a single session. What you described actually sounds like an issue that recently worked with the LV dev team to debug; it was a crash when opening a large number of VIs contained in PPLs that called into other PPLs, and so on. On my end, the number of references getting opened was huge, and although what you described does sound similar, it's very tough to tell if it's the same thing. But if it is, then the 2-pass build could be the only workaround for now.

Regarding the circular dependency, do you have a build log that you could look at to see what Solution Builder thinks is the issue (by giving the tool a log file path)? I'm also curious if you'd consider trying to use one of my other projects [LVDependencyInspector](https://github.com/jovianarts/LVDependencyInspector) to see if that tool also detects a circular dependency? Word of caution, I haven't opened that tool in a while, and I don't recall its functioning state.

Otherwise, is there anything you can think of that I could try on my end to try and help you?

**Comment by labview-mike** [comment]
@jovianarts 
Thanks for the response.

Re: "I've gotten around this by building in two steps" --> I'm just doing this manually. I build everything else, and then I have separate projects for the latest two PPLs.

Re: "circular dependency" --> I don't have a log. I can try using LVDependencyInspector later this week.

If you could point me to the part of the code where the dependencies are being determined, that would be helpful, as I could poke around there.

**Reply by jovianarts** [reply]
@labview-mike , absolutely. The top-level VI that computes dependencies and order is `CalculateSpecificationBuildOrder.vi`. Now the VI that reports the circular dependency is one of that VI's subVIs called `subReduceBuildOrder.vi`. You'll see that the large case structure's `False` case has the error string in it. Good luck!

**Comment by neilfenstein** [comment]
@jovianarts 
I love this app and it should come included in LabVIEW :) 
I was using this successfully and it has been great. But several things have happened and I am struggling to figure out where it has gone wrong. I have changed machines and am tracking down issues. 
It appears that where ever I run the tests some of them fail. I am following the instructions but I am getting failures still.
I believe this has always been the case, but thought I should mention it just in case it helps you or you could advise me.
I am using LV 2020 SP1 and SP1f1 same with both.
Let me know if you need more info or I can do anything to try and resolve.
